File: python/packages/runner/src/varar_runner/results.py
"""results.py — persist run results for the language server (ADR 0014).

The shell half of the run-result contract: the core builds the payload, this
writes ``<root>/.varar/<oath_path>.json`` so the (language-neutral) LSP can turn
a failure into an editor diagnostic. Port of the TypeScript vitest reporter's
writing half; every adapter in this port feeds the same collector, so pytest and
unittest cannot drift from each other.
"""
from __future__ import annotations

import json
from pathlib import Path

from varar_core.hash import hash_source
from varar_core.result import ExampleResult, OathResults, to_wire


def result_file_path(root: Path, oath_path: str) -> Path:
    """``<root>/.varar/<oath_path>.json`` — the file the LSP watches."""
    return root / ".varar" / f"{oath_path}.json"


def write_oath_results(root: Path, results: OathResults) -> Path:
    """Write one oath's results. 2-space indent + trailing newline, matching
    ``JSON.stringify(results, null, 2)`` in the TypeScript port byte-for-byte —
    including leaving non-ASCII unescaped."""
    out = result_file_path(root, results.oath_path)
    out.parent.mkdir(parents=True, exist_ok=True)
    out.write_text(
        json.dumps(to_wire(results), indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    return out


def _document_order(result: ExampleResult) -> tuple[int, str]:
    """Sort key putting examples in document order.

    A test framework reports examples in ITS order — unittest sorts by method
    name, minitest randomises, cargo runs in parallel — so the order results are
    recorded in is not the order they appear in the oath. The file is a
    cross-port contract read by tools that diff runs, so it is written in
    document order everywhere. Name breaks ties for examples sharing a line.
    """
    return (result.lines[0] if result.lines else 0, result.name)


class ResultsCollector:
    """Accumulates each oath's example results across a run, then writes them.

    A test framework hands back one example at a time and only says "the run is
    over" at the end, so the results for an oath cannot be written until then.
    Passing oaths are written too — a stale file would otherwise keep an
    already-fixed diagnostic on screen.
    """

    def __init__(self) -> None:
        self._sources: dict[str, str] = {}
        self._examples: dict[str, list[ExampleResult]] = {}

    def record(self, oath_path: str, source: str, result: ExampleResult) -> None:
        self._sources[oath_path] = source
        self._examples.setdefault(oath_path, []).append(result)

    def write_all(self, root: Path) -> list[Path]:
        written = []
        for oath_path, examples in self._examples.items():
            results = OathResults(
                version=1,
                oath_path=oath_path,
                source_hash=hash_source(self._sources[oath_path]),
                examples=tuple(sorted(examples, key=_document_order)),
            )
            written.append(write_oath_results(root, results))
        return written

File: dotnet/Varar.Runner/Results.cs
using System.Collections.Immutable;
using Varar.Core;

namespace Varar.Runner;

/// <summary>
/// Persists run results for the language server (ADR 0014) — the shell half of the contract the
/// core builds the payload for. Writes <c>&lt;root&gt;/.varar/&lt;oathPath&gt;.json</c>, which the
/// (language-neutral) LSP reads to turn a failure into an editor diagnostic.
/// </summary>
/// <remarks>
/// Lives in the runner so every adapter in this port feeds the same collector and cannot drift
/// from the TypeScript reporter this is a port of.
/// </remarks>
public sealed class Results
{
    private readonly Dictionary<string, string> sources = new(StringComparer.Ordinal);
    private readonly Dictionary<string, List<ExampleResult>> examples = new(StringComparer.Ordinal);

    // Per oath: the other documents its steps were spliced in from (ADR 0016), as path → source.
    private readonly Dictionary<string, Dictionary<string, string>> documents =
        new(StringComparer.Ordinal);

    /// <summary><c>&lt;root&gt;/.varar/&lt;oathPath&gt;.json</c> — the file the LSP watches.</summary>
    public static string ResultFilePath(string root, string oathPath) =>
        Path.Combine(root, ".varar", oathPath.Replace('/', Path.DirectorySeparatorChar) + ".json");

    /// <summary>
    /// Writes one oath's results: 2-space indent plus a trailing newline, matching
    /// <c>JSON.stringify(results, null, 2)</c> in the TypeScript port.
    /// </summary>
    public static string Write(string root, OathResults results)
    {
        var out_ = ResultFilePath(root, results.OathPath);
        Directory.CreateDirectory(Path.GetDirectoryName(out_)!);
        File.WriteAllText(out_, ResultJson.ToWireJson(results) + "\n");
        return out_;
    }

    /// <summary>
    /// Accumulates one example's outcome; the oath is written once its examples are in.
    /// <paramref name="referencedSources"/> carries the OTHER documents this oath's steps were
    /// spliced in from (ADR 0016), as path → source; their hashes go in the payload so a consumer
    /// can tell a stale failure from a live one.
    /// </summary>
    public void Record(
        string oathPath,
        string source,
        ExampleResult result,
        IReadOnlyDictionary<string, string>? referencedSources = null)
    {
        sources[oathPath] = source;
        if (referencedSources is { Count: > 0 })
        {
            if (!documents.TryGetValue(oathPath, out var docs))
            {
                docs = [];
                documents[oathPath] = docs;
            }

            foreach (var (path, text) in referencedSources)
            {
                docs[path] = text;
            }
        }

        if (!examples.TryGetValue(oathPath, out var recorded))
        {
            recorded = [];
            examples[oathPath] = recorded;
        }

        recorded.Add(result);
    }

    /// <summary>
    /// Writes every oath held, and forgets them. Passing oaths are written too — a stale file
    /// would keep a diagnostic on screen that the run has just cleared.
    /// </summary>
    public void FlushAll(string root)
    {
        foreach (var (oathPath, recorded) in examples.ToList())
        {
            var docs = documents.TryGetValue(oathPath, out var held)
                ? held.OrderBy(d => d.Key, StringComparer.Ordinal)
                    .Select(d => new ReferencedDocument(d.Key, Hash.HashSource(d.Value)))
                    .ToImmutableArray()
                : [];
            Write(
                root,
                new OathResults(
                    2,
                    oathPath,
                    Hash.HashSource(sources[oathPath]),
                    [.. recorded],
                    docs));
        }

        examples.Clear();
        documents.Clear();
    }
}
